### DISCUSSION TITLE
Copy state array

### DISCUSSION THREAD

**Discussion by PaganViking** [pmndrs/zustand#1153] [Q&A] [answered]
When I set the state in setCatsTop it sets the state of the useTopCategories store as well.
How can I prevent that ? 
Is there a spesial way to copy the state of the store ?

-------------------------------------------------------------------------------------------

import {useTopCategories,} from "@/hooks/DynamickProps"

const CategoryPanel = () => {
  const categoriesTop = useTopCategories(state => state.categoriesTop)
  
  const [catsTop, setCatsTop] = useState(...[categoriesTop])

  const changeTopCatPublished = (i) => {
    let tempArray = catsTop
    const x = Number(tempArray[i].cPublished)
    tempArray[i].cPublished = x === 1 ? 0 : 1
    setCatsTop([...tempArray])      
  }

return (
{catsTop.map((cat,i) => (
    <>
    <p>{cat.cName}</p>
    <Box
        as="button"
        background={cat.cPublished === 0 ? butBgColor:butBgColorSel}
        borderColor={cat.cPublished === 0 ? butBorderColor:butBorderColorSel}
        color={butTextColor}
        borderWidth="1px"
        borderLeftStyle="solid"
        padding="0px" 
        width="50%" 
        height="32px"
        textAlign="center"
        outline="none"
        cursor= "pointer"
        _hover={{
        borderWidth:"1px",
        borderStyle:"solid",
        borderColor:butBorderColorSel
        }}
        onClick={e => changeTopCatPublished(i)} 
        >
        <span className="ButtonLabel" style={cat.cPublished === 0 ? {color:butTextColor}:{color:butTextColorSel}}>Publisert</span>
    </Box>
    </>
))}
)

}

**Comment by dai-shi** [comment, marked as answer]
You need to avoid mutating the state:
```js
const CategoryPanel = () => {
  const categoriesTop = useTopCategories(state => state.categoriesTop)

  const [catsTop, setCatsTop] = useState(categoriesTop)

  const changeTopCatPublished = (i) => {
    const tempArray = [...catsTop]
    const x = Number(tempArray[i].cPublished)
    tempArray[i] = { ...tempArray[i], cPublished: x === 1 ? 0 : 1 }
    setCatsTop(tempArray)
  }
```

**Reply by PaganViking** [reply]
Thank you :-)
